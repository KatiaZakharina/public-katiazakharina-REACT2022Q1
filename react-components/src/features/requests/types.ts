export type TourFormData = {
  firstName: string;
  lastName: string;
  email: string;
  date: string;
  destination: string;
  withChildren: boolean;
  pcr: File[];
  getNotification: boolean;
};

export type RequestsState = {
  requests: TourFormData[];
};
