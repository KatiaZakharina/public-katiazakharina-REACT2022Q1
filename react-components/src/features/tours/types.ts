import { TourData } from 'services/ToursDataType';

export const SortOptions = {
  'best seller': 'BEST_SELLER',
  'distance from landmark': 'DISTANCE_FROM_LANDMARK',
  'guest rating': 'GUEST_RATING',
  'price highest first': 'PRICE_HIGHEST_FIRST',
  price: 'PRICE',
};

export type FilterData = {
  pageSize: string;
  rating: string;
  sortOrder: typeof SortOptions[keyof typeof SortOptions];
};

type Pagination = { current: number; total: number };

export type TourState = {
  searchValue: string;
  filters: FilterData;
  pagination: Pagination;
  tours: TourData[];
};
