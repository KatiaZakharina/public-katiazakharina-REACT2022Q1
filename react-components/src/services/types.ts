export type CustomErrorCode = 'empty_search';

export interface RequestErrorType extends Error {
  code?: number | CustomErrorCode;
}

export type TourData = {
  id: string;
  accommodation: string;
  rating?: number;
  city: string;
  country: string;
  img: string;
  price?: number;
  landmarks?: string;
};

export interface HotelI {
  id: number;
  name?: string;
  starRating?: number;
  address?: {
    locality?: string;
    countryName?: string;
  };
  landmarks?: Array<{
    label: string;
    distance: string;
  }>;
  ratePlan?: {
    price?: {
      exactCurrent?: number;
    };
  };
  optimizedThumbUrls?: {
    srpDesktop?: string;
  };
}

type TourDescriptionAmenities = {
  heading: string;
  listItems: Array<{
    heading: string;
    listItems: string[];
  }>;
};

export type TourDescriptionType = {
  name: string;
  rating: number;
  price: string;
  address: string;
  tagline: string[];
  guestReviews: string;
  neighborhood: string;
  amenities: TourDescriptionAmenities[];
  optionalExtras: string[];
  arrivingLeaving: string[];
  img?: string;
};

export type TourDetailsType = TourDescriptionType & { img: string };
