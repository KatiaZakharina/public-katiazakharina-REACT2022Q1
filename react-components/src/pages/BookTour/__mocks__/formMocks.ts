import userEvent from '@testing-library/user-event';

import { fields } from '../Form/Form';
import { TourFormData } from 'features/requests/types';

type InputsEls = { [key in keyof TourFormData]: HTMLInputElement };

export const fakeImage = [new File(['image'], 'image.png', { type: 'image/png' })];

export function getInputs(container: HTMLElement): InputsEls {
  return fields.reduce((acc, curr) => {
    const input = container.querySelector(`[name=${curr}]`) as HTMLInputElement;

    if (input) {
      acc[curr] = input;
    }
    return acc;
  }, {} as InputsEls);
}

export async function fillInputs(data: TourFormData, inputs: InputsEls) {
  userEvent.type(inputs.firstName, data.firstName);
  userEvent.type(inputs.lastName, data.lastName);
  userEvent.type(inputs.email, data.email);
  userEvent.type(inputs.date, data.date);

  userEvent.selectOptions(inputs.destination, data.destination);

  const file = data.pcr;
  userEvent.upload(inputs.pcr, file);

  if (data.getNotification) userEvent.click(inputs.getNotification);
  if (data.withChildren) userEvent.click(inputs.withChildren);
}
