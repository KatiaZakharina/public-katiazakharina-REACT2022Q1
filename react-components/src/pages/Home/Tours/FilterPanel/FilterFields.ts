import { Fields } from 'components/Inputs/InputTypes';
import { FilterData, SortOptions } from 'store/reducers/tours/types';

export const inputFields: Fields<HTMLInputElement | HTMLSelectElement, FilterData> = {
  pageSize: {
    id: 'pageSize',
    type: 'number',
    labelText: 'Cards per page',
    max: 25,
    min: 1,
  },
  rating: {
    id: 'rating',
    type: 'number',
    labelText: 'Hotel rating',
    min: 1,
    max: 5,
  },
  sortOrder: {
    id: 'sortOrder',
    labelText: 'Sort by',
    type: 'select',
    options: ['PRICE', ...Object.keys(SortOptions)],
    values: ['PRICE', ...Object.values(SortOptions)],
  },
};
