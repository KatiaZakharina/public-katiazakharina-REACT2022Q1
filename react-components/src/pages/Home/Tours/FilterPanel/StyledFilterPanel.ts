import styled from 'styled-components';

export const StyledFilterPanel = styled.form.attrs({ name: 'filter-panel' })`
  display: flex;
  align-items: center;
  justify-content: space-evenly;
  width: 100%;
`;
