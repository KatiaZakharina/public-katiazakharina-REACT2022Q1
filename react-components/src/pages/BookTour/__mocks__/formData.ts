import { formatYmd } from 'services/dateFormatter';
import { TourFormData } from '../Form/FormFields';
import { fakeImage } from './formMocks';

export const validData: TourFormData = {
  firstName: 'Ivan',
  lastName: 'Ivanov',
  email: 'ivan@gmail.com',
  date: formatYmd(new Date()),
  destination: 'France',
  withChildren: false,
  pcr: fakeImage,
  getNotification: true,
};

export const invalidData: TourFormData = {
  firstName: 'Iva124',
  lastName: '^&8f#',
  email: 'ivan@gmail.com',
  date: '2022-04-06',
  destination: '',
  withChildren: false,
  pcr: [],
  getNotification: true,
};
