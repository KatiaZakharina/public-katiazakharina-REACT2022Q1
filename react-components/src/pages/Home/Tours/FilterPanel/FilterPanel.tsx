import { ChangeEvent } from 'react';

import { customInputs, CustomInputs, Input } from 'components/Inputs';
import { Label } from 'components/Inputs/StyledLabel';
import { inputFields } from './FilterFields';
import { StyledFilterPanel } from './StyledFilterPanel';
import { useAppDispatch, useAppSelector } from 'app/hooks';
import { FilterData } from 'features/tours/types';
import { changeFilters } from 'features/tours/actions';

export const FilterPanel = () => {
  const filters = useAppSelector((state) => state.toursReducer.filters);
  const dispatch = useAppDispatch();

  const fields: Array<keyof FilterData> = ['pageSize', 'rating', 'sortOrder'];

  const onChange = (e: ChangeEvent<HTMLInputElement> & ChangeEvent<HTMLSelectElement>) => {
    dispatch(changeFilters({ ...filters, [e.target.name]: e.target.value }));
  };

  return (
    <StyledFilterPanel>
      {fields.map((name, index) => {
        const type = inputFields[name].type as CustomInputs;
        const InputElement = type in customInputs ? customInputs[type] : Input;

        return (
          <Label htmlFor={name} key={index}>
            {inputFields[name].labelText}
            <InputElement
              name={name}
              inputField={{ ...inputFields[name], value: filters[name] }}
              onChange={onChange}
            />
          </Label>
        );
      })}
    </StyledFilterPanel>
  );
};
