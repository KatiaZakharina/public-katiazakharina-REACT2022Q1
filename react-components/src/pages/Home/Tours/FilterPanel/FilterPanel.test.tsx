import '@testing-library/jest-dom';
import { screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { SortOptions } from 'store/reducers/tours/types';
import { fakeLocalStorage } from 'test/__mocks__/fakeLocalStorage ';
import { customRender } from 'test/__mocks__/customRender';
import { FilterPanel } from './FilterPanel';

describe('FilterPanel', () => {
  beforeAll(() => {
    Object.defineProperty(window, 'localStorage', {
      value: fakeLocalStorage,
    });
  });

  beforeEach(() => {
    window.localStorage.clear();
  });

  it('renders correctly with default values', () => {
    const { asFragment } = customRender(<FilterPanel />);

    expect(asFragment()).toMatchSnapshot();
  });

  it.skip('renders correctly with localStorage values', () => {
    const filtersData = { pageSize: 12, rating: 5, sortOrder: SortOptions['guest rating'] };
    const localData = {
      toursReducer: {
        filters: filtersData,
      },
    };
    window.localStorage.setItem('app', JSON.stringify(localData));

    customRender(<FilterPanel />);

    expect(screen.getByRole('form')).toHaveFormValues(filtersData);
  });

  it.skip('change value and set data in localStorage', async () => {
    const filters = { pageSize: 12, rating: 5, sortOrder: SortOptions['guest rating'] };
    const localStorageFilters = {
      pageSize: '12',
      rating: '5',
      sortOrder: SortOptions['guest rating'],
    };

    customRender(<FilterPanel />);

    const cardPerPage = screen.getByRole('spinbutton', { name: 'Cards per page' });
    const rating = screen.getByRole('spinbutton', { name: 'Hotel rating' });
    const sort = screen.getByRole('combobox', { name: 'Sort by' });

    userEvent.clear(cardPerPage);
    userEvent.clear(rating);

    userEvent.type(cardPerPage, filters.pageSize.toString());
    userEvent.type(rating, filters.rating.toString());
    userEvent.selectOptions(sort, filters.sortOrder);

    expect(screen.getByRole('form')).toHaveFormValues(filters);

    const localData = JSON.parse(window.localStorage.getItem('app') || '');
    expect(localData?.toursReducer?.filters).toStrictEqual(localStorageFilters);
  });
});
