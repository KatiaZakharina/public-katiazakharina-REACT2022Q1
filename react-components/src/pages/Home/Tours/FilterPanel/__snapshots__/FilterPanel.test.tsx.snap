// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterPanel renders correctly with default values 1`] = `
<DocumentFragment>
  <form
    class="sc-jrQzAO abGKM"
    name="filter-panel"
  >
    <label
      class="sc-pVTFL jiYIUu"
      for="pageSize"
    >
      Cards per page
      <input
        class="sc-eCImPb vLPjS"
        id="pageSize"
        max="25"
        min="1"
        name="pageSize"
        type="number"
        value="24"
      />
    </label>
    <label
      class="sc-pVTFL jiYIUu"
      for="rating"
    >
      Hotel rating
      <input
        class="sc-eCImPb vLPjS"
        id="rating"
        max="5"
        min="1"
        name="rating"
        type="number"
        value=""
      />
    </label>
    <label
      class="sc-pVTFL jiYIUu"
      for="sortOrder"
    >
      Sort by
      <select
        class="sc-eCImPb vLPjS"
        id="sortOrder"
        name="sortOrder"
        type="select"
      >
        <option
          hidden=""
          value=""
        >
          PRICE
        </option>
        <option
          value="PRICE"
        >
          best seller
        </option>
        <option
          value="BEST_SELLER"
        >
          distance from landmark
        </option>
        <option
          value="DISTANCE_FROM_LANDMARK"
        >
          guest rating
        </option>
        <option
          value="GUEST_RATING"
        >
          price highest first
        </option>
        <option
          value="PRICE_HIGHEST_FIRST"
        >
          price
        </option>
      </select>
    </label>
  </form>
</DocumentFragment>
`;
